src/common/CryptoUtils.ts

```diff
@@ -0,0 +1,155 @@
+
+export type PEMCertificate = {
+    derCertificate: string;
+    pemCertificate: string;
+    pemPrivateKey: string;
+    pemPublicKey: string;
+};
+
+export class CryptoUtils {
+    /**
+     * Generates a self-signed certificated using the provided parameters.
+     *
+     * @param {string} hostname the hostname to use. Defaults to LOCALHOST.
+     * @param {number} keySize the size for the private key in bits, which must
+     * be between 2048 and 16384. Defaults to 4096.
+     * @param {number | Date} validity the validity length for the certificate.
+     * It can either be a number (representing the number of days from today) or
+     * it can be a specific date (which has to be at least 24 hours from now).
+     * Defaults to 365 days. The maximum length cannot be more than 825 days in the
+     * future to meet Apple guidelines: https://support.apple.com/en-gb/103769
+     * @returns the generated certificate and key.
+     */
+    public static generateSelfSignedCert(
+        hostname: string = 'localhost',
+        keySize: number = 4096,
+        validity: number | Date = 365
+    ): PEMCertificate {
+        if (keySize < 2048 || keySize > 16_384) {
+            throw new Error(messages.getMessage('error:invalidKeySize'));
+        }
+
+        if (validity instanceof Date) {
+            const millisecondsInOneDay = 24 * 60 * 60 * 1000;
+            const diff = validity.getTime() - new Date().getTime();
+            if (diff < millisecondsInOneDay || diff > 825 * millisecondsInOneDay) {
+                throw new Error(messages.getMessage('error:invalidValidityDate'));
+            }
+        } else if (validity < 1 || validity > 825) {
+            throw new Error(messages.getMessage('error:invalidValidityNumber'));
+        }
+
+        const keys = forge.pki.rsa.generateKeyPair(keySize);
+        const cert = forge.pki.createCertificate();
+
+        const startDate = new Date();
+        let endDate = new Date();
+        if (validity instanceof Date) {
+            endDate = validity;
+        } else {
+            endDate.setDate(startDate.getDate() + validity);
+        }
+
+        cert.publicKey = keys.publicKey;
+        cert.serialNumber = '01';
+        cert.validity.notBefore = startDate;
+        cert.validity.notAfter = endDate;
+
+        const attrs = [
+            { name: 'commonName', value: hostname },
+            { name: 'countryName', value: 'US' },
+            { shortName: 'ST', value: 'California' },
+            { name: 'localityName', value: 'San Francisco' },
+            { name: 'organizationName', value: 'Salesforce Inc.' },
+            { shortName: 'OU', value: 'LocalDevPreview' }
+        ];
+
+        cert.setSubject(attrs);
+        cert.setIssuer(attrs);
+
+        // Add the subjectAltName and ExtendedKeyUsage extensions, which are required by Apple
+        cert.setExtensions([
+            {
+                name: 'basicConstraints',
```

Comment:
This isn't a CA certificate. We should either omit this section entirely, or set `CA` to `false`.

Reply:
If I remove this section entirely or if I set `CA = false` then after installing the cert on my sim and going to `Settings > General > VPN & Device Management > localhost` it shows that the certificate is not verified. But by including this section and having `CA = true`, it would show as verified.

So you're positive that we don't need it?
![Screenshot 2024-05-31 at 4 36 00 PM](https://github.com/forcedotcom/lwc-dev-mobile-core/assets/65030660/115c0610-46c7-4f3c-b906-785ef225ed4d)

Reply:
My experience is a bit apples and oranges, as I'm typically creating these certificates with `openssl`. We can leave it for now, and tweak later as necessary.